### ISSUE TITLE
Discord Rich Presence Memory Leak?

### ISSUE THREAD

**Issue by michaeltcoyle** [MarshallOfSound/Google-Play-Music-Desktop-Player-UNOFFICIAL-#2969]
**OS:** Windows 10 Pro 64 bit

**Issue Descriptions:** Having Discord Rich Presence setting enabled without Discord open results in GPMDP using more and more memory and cpu until application crash.

**Steps to Reproduce:** Have both Discord and GPMDP closed. Open GPMDP and make sure Discord Rich Presence setting is enabled. Play a song and watch task manager. After a few minutes the application will hang.

**Comment by MarshallOfSound** [comment]
@michaeltcoyle Investigating now 👍

**Comment by MarshallOfSound** [comment]
@michaeltcoyle I can't reproduce this behavior, I see higher than expected CPU usage from GPMDP while discord is closed (averaging just below 8%), that is probably due to aggressive attempts to reconnect to discord.

Can you screen cast your resource usage increase?

**Comment by Alipoodle** [comment]
Cannot reproduce. Windows 10, 64bit.
Even tested with Discord open while not connected (Had troubles with other applications like this before)

**Comment by Alipoodle** [comment]
So while making issue #2970 I had been looking through console of Discord to see which presences was coming through and info... and GPMDP seemed to be emitting WAY to many RichPresence requests than it should be doing. [Video](https://puu.sh/zdwxB/1b0568acae.mp4)

GPMDP Should only trigger an update in Rich Presence in 3 cases, and most likely one of these events are triggering too many times or something is wrong within the code to prevent it from spamming it.
```
PlaybackAPI.on('change:state', setPresence);
PlaybackAPI.on('change:track', setPresence);
PlaybackAPI.on('change:time', setPresence);
```

**Comment by MarshallOfSound** [comment]
@Alipoodle `change:time` triggers more as a "progress" event and triggers many times a second, is this not the expected use case of the duration property?

**Comment by G4Zz0L1** [comment]
I've the same problem, on Mint with the Linux client. CPU 100% and increasing ram until crash.

![image](https://user-images.githubusercontent.com/13220636/35620594-e1d9c6a2-0681-11e8-9b90-7c0bc6f59bf9.png)

**Comment by MarshallOfSound** [comment]
@G4Zz0L1 And this doesn't happen if discord rich presence is disabled?

**Comment by G4Zz0L1** [comment]
I've killed the instance and disabled that setting, now it's like that

![image](https://user-images.githubusercontent.com/13220636/35620849-add674f8-0682-11e8-8ae2-c4453b66152f.png)

**Comment by MarshallOfSound** [comment]
@G4Zz0L1 Just to double check you chose to enable that setting right?  I didn't screw up and enable it by default?

**Comment by G4Zz0L1** [comment]
Yeah, I've enabled it for test, it's not your fault   xD

**Comment by MarshallOfSound** [comment]
@G4Zz0L1 Sweet, I'll look into the weird CPU / memory usage but for now at least it doesn't affect everyone 😄

**Comment by G4Zz0L1** [comment]
If you need something, ask and I'll provide asap

**Comment by greesymnky** [comment]
I am having this issue since updating to latest version yesterday. I can use the GPDMP once it starts but after a few minutes my computer is in overdrive and the program becomes extremely laggy until it doesn't respond. I have deselected the Discord Rich Presence Integration setting and that has seemed to eliminate the issue so far.

**Comment by MarshallOfSound** [comment]
@greesymnky And all people in this issue (or just arriving) can you all please confirm in your comments that are effectively "me too" that you explicitly enabled the Discord Rich Presence integration.  It isn't clear in some peoples comments whether something broke and auto-enabled it or not

**Comment by michaeltcoyle** [comment]
Yes I narrowed it down to being an issue with the Rich presence setting for me.

**Comment by dblooom** [comment]
On my Mac, I have also memory problems after last update. Eats above 100% CPU 
![bildschirmfoto 2018-02-01 um 09 51 50](https://user-images.githubusercontent.com/1446968/35669627-03d755d0-0736-11e8-9137-514c343e95cc.png)

**Comment by dblooom** [comment]
Hm. Now after restarting the app, CPU eat seems normal again

**Comment by Bryzzle** [comment]
7% CPU usage, 1.2GB of ram used. Windows 10. Fixed after disabling Discord Rich Presence

**Comment by jostrander** [comment]
According to the rich presence docks the maximum update speed is every 15
seconds. The Implementation here has multiple events at the millisecond
level being fired which is probably causing the memory to spike as the
events stack up.

**Comment by symphonicon** [comment]
I'm having this issue too. I turned on Discord Rich Presence and in a few minutes application becomes extremely lagyy to the point when it doesn't react, but still plays music. With DRP on CPU usage is about 50-60% and RAM at least 1.5 GB. With DRP turned off CPU is still at 50%, but RAM gets from 500 MB and rising up constantly, every second for a few megabytes. 

_Win10 64bit, Intel Core i5, 12GB RAM._

**Comment by vanushwashere** [comment]
Same on Ubuntu 17.10
:(

**Comment by fsenthus** [comment]
I also am experiencing this. Was trying to listen to a radio station and after a few songs would play, GPMDP started to experience major UI lag, where it took several seconds to recognize any button press, etc... couldn't maximize the window or anything. Task manager showed it was using 1300 MB and climbing RAM. Disabled Discord rich presence setting and it has so far solved it.

**Comment by jaccarmac** [comment]
I seem to be experiencing this on Gentoo. Upgraded GPDMP and turned on rich presence, after playing music for a while the app completely freezes and has to be force closed. While frozen it still plays audio. It also crashed on me a few days ago completely after a couple hours. Possibly related but I'm not sure.

**Comment by shadowsparkyfox** [comment]
Yep yep, had a leaky leak. 
1.5GB of RAM
Confirmed it like this: 
Turned on Discord Rich presence without Discord running. The app started hogging more and more memory. Closed it to the system tray and it stopped grabbing memory, the moment I reopened it it started hogging again.

From what I notice it only sucks up memory if the GUI is open. If it is in the tray it doesn't do the sucky suck. ;)

**Comment by shadowsparkyfox** [comment]
Googled the Github repo and found this issue in a few moments ;)
besides this bug great stuff by the way @MarshallOfSound (y)

**Comment by horusxci** [comment]
Had the same issue, Launched the app without discord open and it slowly starts stacking ram untill it becomes unresponsive, no issue when discord is already open at the moment of launching it or if rich integration is disabled. Also no issue if you close discord after the app has been opened.

![image](https://user-images.githubusercontent.com/37197748/37184082-101c9df2-22ff-11e8-8bce-45412e59703d.png)

**Comment by cmcpasserby** [comment]
yeah am having the same issue, disabling discord integration fixed it for me in the mean time. But on Windows i was reaching 1.3gb's of memory to used and a very slow and non responsive interface in GMDP

**Comment by fraserhaer** [comment]
@jostrander @MarshallOfSound 

I saw in the pull request neither of you were experiencing the issue but thought you'd found the fix. I'm now using the update from AppVeyor built 17 hours ago and can confirm that fixed it for me.

**Comment by scj643** [comment]
Can confirm on Fedora Workstation 27 as well

**Comment by XZaya** [comment]
Reproduced at Windows 10 64bit!
CPU and Memory was as 70%, and I couldn't even start a Song without waiting 10 Seconds.

**Comment by Zaf9670** [comment]
Windows 10 Pro 64-bit. Had the same CPU and Memory usage creeping up after 10-20 minutes of play. Disabled Discord options and Windows 10 integration and it went back to normal. Going to try with W10 integration again to make sure it's isolated to Discord.

**Comment by kpuljek** [comment]
Confirmed, came to report memory leak (Ubuntu 16.04) and I just today switched Discord back to browser, rich presence was enabled.

**Comment by GameKyuubi** [comment]
Just ran into this on Windows 10.

**Comment by Alipoodle** [comment]
**OK** it's known that it is broken! >.<

 Can we please not have more that reports of "can reproduce" and have something else that will benefit this issue!

**Comment by Alipoodle** [comment]
Fixed and Can Close!

**Comment by Kintar** [comment]
I'm using v4.6.1 and this is still occurring.  I see that @Alipoodle commented that the issue was fixed, but I don't see a pull request and this ticket is still open, now almost a year later (2019-03-01).  Has this fallen off the radar, or was it not really fixed?  I'm happy to look into it, but don't want to duplicate effort if it really is fixed and just not merged.

**Comment by Alipoodle** [comment]
Probably should have put the pull request regarding this.

This was "meant" to have been fixed in #3001. *Summaries, it basically delays the `change:time` event to only submit every 15 seconds rather than each time the event triggers (which is a lot)*

There might possibly be issues to with Mac specifically now.

**Comment by Kintar** [comment]
Gotcha.  I'll pull down the source and do a little digging to see what's causing it on a Mac.

**Comment by JoeJoeTV** [comment]
I am using Version 4.6.1 of GPMDP on Linux Min 19.1 64 bit and I can confirm that this only happening, when discord rich presence is turned on.
